File: 214/projects/08/java/Temp.java
/* Temp.java implements class Name.
 * Prof: Dr. Adams, for CS 214 at Calvin College.
 * Student: Chan Kim (ck45)
 * Date: 4/9/2018
 ************************************************************/

import java.io.*;
import java.util.Scanner;

class Temp {
  
    /*
     * Temp(degrees, scale) initializes a Temp variable          
     * Receive: degrees, the degrees ;         
     *          scale, the scale .         
     * Return: set to degrees, and scale if it is valid 
      *        otherwise prints an invalid message.    
     */
    public Temp(double degrees, char scale) {

        char capScale = Character.toUpperCase(scale);

        if (isValid(degrees, capScale))
        {
            myDegrees = degrees;
            myScale = capScale;
        }
        else
        {
            System.out.println("Invalid value");
        }
    }

    /*
     * isValid(degrees, scale) checks the validation.        
     * Receive: degrees, float; scale, a character.                      
     * PRE: degrees and scale have been initialized.            
     * Return: true or false.     
     */

    public boolean isValid(double degrees, char scale)
    {
        if (scale == 'F')
        {
            if (degrees < -459.67)
            {
                return false;
            }
            else
            {
                return true;
            }
        }
        else if (scale == 'C')
        {
            if (degrees < -273.15)
            {
                return false;
            }
            else
            {
                return true;
            }
        }
        else if (scale == 'K')
        {
            if (degrees < 0.0)
            {
                return false;
            }
            else
            {
                return true;
            }
        }
        else
        {
            return false;
        }                 
    }

    // extractor for first name of a Name object
    public  double getDegrees() {
        return myDegrees;
    }

    // extractor for middle name of a Name object
    public  char getScale() {
        return myScale;
    }

    /*
     * isValid(degrees, scale) checks the validation.        
     * Receive: degrees, float; scale, a character.                      
     * PRE: degrees and scale have been initialized.            
     * Return: true or false.
     */

    public void setFahrenheit()
    {
        if (getScale() == 'C')
        {
            myDegrees = ((getDegrees() * 1.8) + 32.0);
            myScale = 'F';
        }
        else if (getScale() == 'K')
        {
            myDegrees = ((getDegrees() * 1.8) - 459.67);
            myScale = 'F';
        }
    }

    public void setCelsius()
    {
        if (getScale() == 'F')
        {
            myDegrees = ((getDegrees() - 32.0) / 1.8);
            myScale = 'C';
        }
        else if (getScale() == 'K')
        {
            myDegrees = (getDegrees() - 273.15);
            myScale = 'C';
        }
    }

    public void setKelvin()
    {
        if (getScale() == 'C')
        {
            myDegrees = (getDegrees() + 273.15);
            myScale = 'K';
        }
        else if (getScale() == 'F')
        {
            myDegrees = ((getDegrees() + 459.67) * (5.0/9.0));
            myScale = 'K';
        }
    }

    /*
     * read(keyboard) reads degrees and scale   
     * Receive: keyboard, Scanner.
     * PRE: keyboard ahs been initialized.                                  
     * Return: sets up degrees and scale if it is valid 
     *          otherwise prints an invalid message.    
    */

    public void read(Scanner keyboard)
    {
        String values = keyboard.nextLine();
        String [] parts = values.split(" ");
        double degrees = Double.parseDouble(parts[0]);
        char scale = parts[1].charAt(0);
        char capScale = Character.toUpperCase(scale);
        if (isValid(degrees, capScale))
        {
            myDegrees = degrees;
            myScale = capScale;
        }
        else
        {
            System.out.println("Invalid value");
        }    
    }

    /*
     * raise(degree) incriments degrees                  
     * Receive: degree, number.                     
     * PRE: degrees has been initialized.             
     * Return: incriments degrees otherwise prints an invalid message.    
    */

    public void raise(double degrees)
    {
        myDegrees = myDegrees + degrees;
    }

    /*
     * lower(degree) decriments degrees                  
     * Receive: degree, number.                     
     * PRE: degrees has been initialized.               
     * Return: decriments degrees otherwise prints an invalid message. 
    */

    public void lower(double degrees)
    {
        double Degrees = myDegrees - degrees;
        if (isValid(Degrees, myScale))
        {
            myDegrees = Degrees;
        }
        else
        {
            System.out.println("Invalid value");
        }          
    }

    /*
     * equals(temp) retrieves baseTemp = limitTemp        
     * Receive: temp, a Temp.                      
     * PRE: baseTemp and limitTemp have been initialized.            
     * Return: true or false. 
    */
    public boolean equals(Temp temp)
    {
        temp.setCelsius();
        setCelsius();
        if (temp.getDegrees() == getDegrees())
        {
            return true;
        }
        else
        {
            return false;
        }
    }

    /*
     * less_than(temp) retrieves baseTemp < limitTemp        
     * Receive: temp, a Temp.                      
     * PRE: baseTemp and limitTemp have been initialized.            
     * Return: true or false.     
    */

    public boolean less_than(Temp temp)
    {
        temp.setCelsius();
        setCelsius();
        if (temp.getDegrees() > getDegrees())
        {
            return true;
        }
        else
        {
            return false;
        }        
    }    

    /*
     * toString() retrieves Temp as a String                                       
     * PRE: temp has been initialized.           
     * Return: a String representation of Temp.      
    */

    public String toString() {
        return Double.toString(myDegrees) + ' ' + Character.toString(myScale);
    }

    // private instance variables
    private double myDegrees;
    private char myScale;  
    }

File: 214/projects/08/ruby/Temp.rb
# Temp.rb implements class Temp.
#
# Prof: Dr. Adams, for CS 214 at Calvin College.
# Student: Chan Kim (ck45)
# Date: 4/9/2018
####################################################

module Temps

    class Temp

        ####################################################
        # initialize(degrees, scale) initializes a Temp variable          
        # Receive: degrees, the degrees ;         
        #          scale, the scale .         
        # Return: set to degrees, and scale if it is valid.
        ####################################################

        def initialize(degrees, scale)
            if (isValid(degrees, scale.upcase)) then
                @degrees, @scale = degrees, scale.upcase
            else
                print "Invalid value"
            end
        end

        attr_reader :degrees, :scale

        ####################################################
        # isValid(degrees, scale) checks the validation.        
        # Receive: degrees, float; scale, a character.                      
        # PRE: degrees and scale have been initialized.            
        # Return: true or false.
        ####################################################

        def isValid(degrees, scale)
            if (scale == 'F') then
                if (degrees < -459.67)
                    return false;
                else
                    return true;
                end
            elsif (scale == 'C') then
                if (degrees < -273.15)
                    return false;
                else
                    return true;
                end
            elsif (scale == 'K') then
                if (degrees < 0.0)
                    return false;
                else
                    return true;
                end
            else
                return false;
            end
        end

        ####################################################
        # setFahrenheit(), setCelsius(), setKelvin()                                  
        # PRE: degrees and scale have been initialized.             
        # Return: sets to those scales    
        ####################################################

        def setFahrenheit()
            if (@scale == 'C')
                @degrees = ((@degrees * 1.8) + 32.0)
                @scale = 'F'
            elsif (@scale == 'K')
                @degrees = ((@degrees * 1.8) - 459.67)
                @scale = 'F'
            end
        end

        def setCelsius()
            if (@scale == 'F')
                @degrees = ((@degrees - 32.0) / 1.8)
                @scale = 'C'
            elsif (@scale == 'K')
                @degrees = (@degrees - 273.15)
                @scale = 'C'
            end
        end

        def setKelvin()
            if (@scale == 'C')
                @degrees = (@degrees + 273.15)
                @scale = 'K'
            elsif (@scale == 'F')
                @degrees = ((@degrees + 459.67) * (5.0/9.0))
                @scale = 'K'
            end
        end

        ####################################################
        # read() reads degrees and scale                                     
        # Return: sets up degrees and scale if it is valid 
        #          otherwise prints an invalid message. 
        ####################################################

        def read()
            values = gets
            parts = values.split(" ")
            degrees = parts[0].to_f
            scale = parts[1].to_s
            if (isValid(degrees, scale.upcase)) then
                @degrees, @scale = degrees, scale.upcase
            else
                print "Invalid value"
            end
        end
        
        ####################################################
        # raiseTemp(degree) incriments degrees                  
        # Receive: degree, number.                     
        # PRE: degrees has been initialized.             
        # Return: incriments degrees otherwise prints an invalid message.
        ####################################################
        
        def raise(degree)
            @degrees = @degrees + degree
        end

        ####################################################
        # lower(degree) decriments degrees                  
        # Receive: degree, number.                     
        # PRE: degrees has been initialized.               
        # Return: decriments degrees otherwise prints an invalid message. 
        ####################################################

        def lower(degree)
            if (isValid((@degrees - degree), @scale)) then
                @degrees = @degrees - degree
            else
                print "Invalid value"
            end            
        end

        ####################################################
        # equals(temp) retrieves baseTemp = limitTemp        
        # Receive: temp, a Temp.                      
        # PRE: baseTemp and limitTemp have been initialized.            
        # Return: true or false. 
        ####################################################

        def equals(temp)
            temp.setCelsius
            setCelsius
            if (temp.degrees == @degrees) then
                return false;
            else
                return true;
            end
        end

        ####################################################
        # less_than(temp) retrieves baseTemp < limitTemp        
        # Receive: temp, a Temp.                      
        # PRE: baseTemp and limitTemp have been initialized.            
        # Return: true or false. 
        ####################################################

        def less_than(temp)
            temp.setCelsius
            setCelsius
            if (temp.degrees > @degrees) then
                return true;
            else
                return false;
            end
        end            
        ####################################################
        # write() writes Temp                                       
        # PRE: degrees has been initialized.           
        # Return: prints Temp.                     
        ####################################################

        def write()
            print @degrees.to_s + " " + @scale
        end
    end
end